File: library/src/main/java/com/github/kongpf8848/rxhttp/request/AbsRequest.java
package com.github.kongpf8848.rxhttp.request;

import com.github.kongpf8848.rxhttp.RxHttp;
import com.github.kongpf8848.rxhttp.callback.HttpCallback;


import java.util.HashMap;
import java.util.Map;

import okhttp3.RequestBody;

public abstract class AbsRequest {

    protected String url;
    protected Object tag;
    protected Map<String,String> headers=new HashMap<>();
    protected Map<String,Object> params=new HashMap<>();

    protected abstract RequestBody buildRequestBody();

    AbsRequest(String url){
        this.url=url;
    }

    public AbsRequest tag(Object tag){
        this.tag=tag;
        return this;
    }
    public AbsRequest header(String key, String value){
        this.headers.put(key,value);
        return this;
    }
    public AbsRequest headers(Map<String,String>headers){
        if(headers!=null) {
            this.headers = headers;
        }
        return this;
    }
    public AbsRequest param(String key, Object value){
        this.params.put(key,value);
        return this;
    }
    public AbsRequest params(Map<String,Object>params){
        if(params!=null){
            this.params=params;
        }
        return this;
    }


    public String getUrl() {
        return url;
    }

    public Object getTag() {
        return tag;
    }

    public Map<String, String> getHeaders() {
        return headers;
    }
    public Map<String, Object> getParams() {
        return params;
    }

    public <T> void enqueue(HttpCallback<T> callback)
    {
        RxHttp.getInstance().enqueue(this,callback);
    }
}

File: RxHttp/src/main/java/io/github/kongpf8848/rxhttp/request/AbsRequest.kt
package io.github.kongpf8848.rxhttp.request

import android.app.Activity
import android.content.Context
import androidx.fragment.app.Fragment
import io.github.kongpf8848.rxhttp.RxHttp
import io.github.kongpf8848.rxhttp.RxHttpTagManager
import io.github.kongpf8848.rxhttp.callback.HttpCallback
import okhttp3.RequestBody

abstract class AbsRequest {

    private var context: Any? = null
    private var url: String = ""
    private var tag: Any? = null
    private var param: Map<String, Any?>? = null
    protected var callback:HttpCallback<*>?=null

    protected abstract fun buildRequestBody(): RequestBody?

    constructor(context: Context) {
        this.context = context
    }

     constructor(fragment: Fragment) {
        context = fragment
    }

     constructor(activity: Activity) {
        context = activity
    }

    fun url(url: String): AbsRequest {
        this.url = url
        return this
    }

    fun tag(tag: Any?): AbsRequest {
        this.tag = tag
        return this
    }

    fun params(params: Map<String, Any?>?): AbsRequest {
        this.param = params
        return this
    }

    val actualContext: Context?
        get() {
            if (context is Context) {
                return context as Context
            } else if (context is Fragment) {
                return (context as Fragment).requireContext()
            }
            return null
        }

    fun getTag(): Any? {
        return if (tag == null) {
            RxHttpTagManager.generateRandomTag()
        } else tag
    }

    fun getContext()=context

    fun getUrl()=url

    fun getParams(): Map<String, Any?>? {
        return param?.filterValues { it!=null }
    }

    fun <T> enqueue(callback: HttpCallback<T>) {
        this.callback=callback
        RxHttp.getInstance().enqueue(this, callback)
    }
}